{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4551735e-a9c9-41b9-8b95-d0a475bba755",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "from pathlib import Path\n",
    "import traceback\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "830321a8",
   "metadata": {},
   "source": [
    "### Main code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc7ee680",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Bowtie2Aligner:\n",
    "    def __init__(self, current_path):\n",
    "        self.bowtie2_index = current_path/\"contaminants_index\"\n",
    "        self.r1_filename = None\n",
    "        self.r2_filename = None\n",
    "        \n",
    "    def single_reads(self, bamfile, file, mapped_folder, unmapped_folder, samtools_folder):\n",
    "        \"\"\"\n",
    "        Align single-end reads (merged/unpaired)\n",
    "        \"\"\"\n",
    "        filename = file.name.split(\".\")[0]\n",
    "        \n",
    "        try:\n",
    "            sam_output = samtools_folder/f\"{filename}_mapped.sam\"\n",
    "            contam_output = mapped_folder/f\"{filename}_mapped.fastq.gz\"\n",
    "            rmcontam_output = unmapped_folder/f\"{filename}_unmapped.fastq.gz\"\n",
    "\n",
    "            cmd = [\"bowtie2\", \n",
    "                    \"-x\", str(self.bowtie2_index), ## provides index that we created earlier\n",
    "                    \"-U\", str(file), ## specifies single-read\n",
    "                    \"--un-gz\", str(rmcontam_output), ## merged/unpaired reads that failed to align (mrna)\n",
    "                    \"--al-gz\", str(contam_output)] ## merged/unpaired reads that aligned ≥1 times\n",
    "            if not bamfile:\n",
    "                cmd.extend([\"-S\", \"/dev/null\"]) ## get empty output\n",
    "            if bamfile:\n",
    "                cmd.extend([\"-S\", str(sam_output)]) ## .sam file of contam reads\n",
    "                \n",
    "            result = subprocess.run(cmd, \n",
    "                                    check = True, \n",
    "                                    capture_output = True, \n",
    "                                    text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to align merged or unpaired fastq file {file.name}: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise\n",
    "        return result\n",
    "\n",
    "    def detect_reps(self, subfolder):\n",
    "        \"\"\"\n",
    "        Defines unmerged R1 and R2 filename variables\n",
    "        \"\"\"\n",
    "        for r1_file in subfolder.glob(\"*_unmerged_R1_*\"):\n",
    "            self.r1_filename = r1_file\n",
    "            self.r2_filename = r1_file.with_name(r1_file.name.replace(\"_R1_\", \"_R2_\"))\n",
    "\n",
    "    def paired_reads(self, bamfile, file, mapped_folder, unmapped_folder, samtools_folder):\n",
    "        \"\"\"\n",
    "        Align paired-end reads (unmerged)\n",
    "        \"\"\"\n",
    "        filename = file.name.split(\".\")[0]\n",
    "\n",
    "        try:\n",
    "            sam_output = samtools_folder/f\"{filename}_mapped.sam\"\n",
    "            contam_output = mapped_folder/f\"{filename}_mapped.fastq.gz\"\n",
    "            rmcontam_output = unmapped_folder/f\"{filename}_unmapped.fastq.gz\"\n",
    "            self.detect_reps(file.parent)\n",
    "\n",
    "            cmd = [\"bowtie2\",\n",
    "                    \"-x\", str(self.bowtie2_index),\n",
    "                    \"-1\", str(self.r1_filename),\n",
    "                    \"-2\", str(self.r2_filename),\n",
    "                    \"--un-conc-gz\", str(rmcontam_output), ## unmerged reads that failed to align (mrna)\n",
    "                    \"--al-conc-gz\", str(contam_output)] ## unmerged reads that aligned ≥1 times\n",
    "            if not bamfile:\n",
    "                cmd.extend([\"-S\", \"/dev/null\"]) ## get empty output\n",
    "            if bamfile:\n",
    "                cmd.extend([\"-S\", str(sam_output)]) ## .sam file of contam reads\n",
    "\n",
    "            result = subprocess.run(cmd, \n",
    "                                    check = True, \n",
    "                                    capture_output = True, \n",
    "                                    text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to align unmerged fastq file {file.name}: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise\n",
    "        return result\n",
    "    \n",
    "    def convert_sam(self, samtools_folder, file):\n",
    "        \"\"\"\n",
    "        Converts .sam output from bowtie2 into .bam\n",
    "        \"\"\"\n",
    "        filename = file.name.split(\".\")[0]\n",
    "        sam_input = samtools_folder/f\"{filename}_mapped.sam\"\n",
    "        sam_filename = Path(sam_input).stem\n",
    "        bam_output = samtools_folder/f\"{sam_filename}_out.bam\"\n",
    "\n",
    "        try:\n",
    "            subprocess.run([\"samtools\", \"sort\", \"-O\", \"BAM\", ## convert .sam to .bam and sort\n",
    "                            \"-o\", str(bam_output), ## output file name\n",
    "                            str(sam_input)], ## input file name\n",
    "                            check = True,\n",
    "                            capture_output = True,\n",
    "                            text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to convert {sam_input.name} to .bam: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise\n",
    "\n",
    "    def merge_bam(self, samtools_folder, file):\n",
    "        \"\"\"\n",
    "        Merges all .bam files, then \n",
    "        sorts and indexes into .bai\n",
    "        \"\"\"\n",
    "        base_name = file.stem.split(\"_\")[0] # create general filestem\n",
    "        merged_bam = samtools_folder/f\"{base_name}_mapped.bam\" # merged output\n",
    "        bam_list = [*samtools_folder.glob(\"*.bam\")] # detect .bam files\n",
    "        rm_list = [*samtools_folder.glob(\"*out.bam\"), *samtools_folder.glob(\"*.sam\")]\n",
    "\n",
    "        try:\n",
    "            subprocess.run([\"samtools\", \"merge\", ## merge all .bam files into one\n",
    "                            str(merged_bam), *map(str, bam_list)], ## asterisk = expand list & iterate through\n",
    "                            check = True, \n",
    "                            capture_output = True,\n",
    "                            text = True)\n",
    "            subprocess.run([\"samtools\", \"index\", str(merged_bam)], ## create .bai from .bam\n",
    "                            check = True,\n",
    "                            capture_output = True,\n",
    "                            text = True)\n",
    "            subprocess.run([\"rm\", *map(str, rm_list)], ## remove original .sam and .bam files\n",
    "                            check = True, ## ensures that this block only runs if previous 2 were successful\n",
    "                            capture_output = True,\n",
    "                            text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to create {merged_bam.name} and convert to .bai: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca57fc6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rmcontam_pipeline(folder_name, bamfile):\n",
    "    ## define input directory\n",
    "    current_path = Path.cwd()\n",
    "    input_dir = current_path/folder_name\n",
    "\n",
    "    ## initialize class\n",
    "    aligner = Bowtie2Aligner(current_path)\n",
    "\n",
    "    for subfolder in input_dir.iterdir(): ## amount of subfolders = number of replicates\n",
    "        if subfolder.is_dir():\n",
    "            mapped_folder = current_path/\"filtered_processed_fastqs\"/\"mapped_contam\"/folder_name/f\"{subfolder.name}\"\n",
    "            mapped_folder.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "            unmapped_folder = current_path/\"filtered_processed_fastqs\"/\"removed_contam\"/folder_name/f\"{subfolder.name}\"\n",
    "            unmapped_folder.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "            samtools_folder = mapped_folder/\"samtools\"\n",
    "            if bamfile:\n",
    "                samtools_folder.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "            for file in subfolder.glob(\"*.fastq.gz\"): ## iterate through indiv. files in subfolder\n",
    "                try:\n",
    "                    ## run bowtie2 alignment functions\n",
    "                    if \"_merged\" in file.name or \"_unpaired\" in file.name:\n",
    "                        aligner.single_reads(bamfile, file, mapped_folder, unmapped_folder, samtools_folder)\n",
    "                    elif \"_unmerged_R1\" in file.name:\n",
    "                        aligner.paired_reads(bamfile, file, mapped_folder, unmapped_folder, samtools_folder)\n",
    "\n",
    "                    ## if user requests it, then a BAM file will be outputted\n",
    "                    if bamfile:\n",
    "                        ## run samtools function\n",
    "                        aligner.convert_sam(samtools_folder, file)\n",
    "                except Exception as e:\n",
    "                    print(f\"Failed to align {file.name} with bowtie2: {e}\")\n",
    "                    traceback.print_exc()\n",
    "                    continue\n",
    "            \n",
    "            if bamfile:\n",
    "                ## merge bam files, convert to bai, & remove files\n",
    "                aligner.merge_bam(samtools_folder, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23b1c93b",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser(description = \"Run contaminant removal pipeline.\")\n",
    "    parser.add_argument(\"--folder_name\", required = True, help = \"Input processed_fastqs folder name\")\n",
    "    parser.add_argument(\"-B\", \"--bamfile\", action = argparse.BooleanOptionalAction, default = False, help = \"Creates BAM file for mapped contaminants; disabled by default\")\n",
    "\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    print(\"Starting contaminant removal pipeline...\")\n",
    "    rmcontam_pipeline(args.folder_name, args.bamfile)\n",
    "    print(\"Pipeline finished.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
