{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4551735e-a9c9-41b9-8b95-d0a475bba755",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import shutil\n",
    "from pathlib import Path\n",
    "import traceback\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "830321a8",
   "metadata": {},
   "source": [
    "### Main code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc7ee680",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Bowtie2Aligner:\n",
    "    def __init__(self, folder_path, file, processed_folder, subfolder):\n",
    "        self.parent_path = Path(folder_path).parent\n",
    "        self.contaminants_dir = self.parent_path/\"contaminants.fa\"\n",
    "        self.bowtie2_index = self.parent_path/\"contaminants_index\"\n",
    "        self.r1_filename = None\n",
    "        self.r2_filename = None\n",
    "        self.file_stem = file.name.split(\".\")[0]\n",
    "        self.sam_output = processed_folder/subfolder/\"samtools\"/f\"{self.file_stem}_mapped.sam\" ## file_stem = og filename w/o both extensions\n",
    "        self.rmcontam_output = processed_folder/subfolder/f\"{self.file_stem}_unmapped.fastq.gz\"\n",
    "        self.contam_output = processed_folder/subfolder/f\"{self.file_stem}_mapped.fastq.gz\"\n",
    "\n",
    "    def build_bowtie2_index(self):\n",
    "        \"\"\"\n",
    "        Must have contaminants.fa in parent directory of folder_path.\n",
    "            folder_path = Folder name that ends with 'processed_fastqs'\n",
    "        \"\"\"\n",
    "        if not self.bowtie2_index.exists():\n",
    "            try:\n",
    "                cmd = [\"bowtie2-build\", \"-f\", \n",
    "                       str(self.contaminants_dir), \n",
    "                       str(self.bowtie2_index)]\n",
    "                result = subprocess.run(cmd, \n",
    "                                        check = True, ## if command returns non-zero exit status, raise error\n",
    "                                        capture_output = True, \n",
    "                                        text = True)\n",
    "            except subprocess.CalledProcessError as e: ## error handling\n",
    "                    print(f\"Failed to build bowtie2 index: {e}\")\n",
    "                    print(\"STDERR:\", e.stderr)\n",
    "                    print(\"STDOUT:\", e.stdout)\n",
    "                    traceback.print_exc()\n",
    "                    raise\n",
    "        else:\n",
    "            pass\n",
    "        return result\n",
    "        \n",
    "    def single_reads(self, file):\n",
    "        \"\"\"\n",
    "        Align single-end reads (merged/unpaired)\n",
    "        \"\"\"\n",
    "        try:\n",
    "            cmd = [\"bowtie2\", \n",
    "                    \"-x\", str(self.bowtie2_index), ## provides index that we created earlier\n",
    "                    \"-U\", str(file), ## specifies single-read\n",
    "                    \"-S\", str(self.sam_output), ## .sam file of contam reads\n",
    "                    \"--un-gz\", str(self.rmcontam_output), ## merged/unpaired reads that failed to align (mrna)\n",
    "                    \"--al-gz\", str(self.contam_output)] ## merged/unpaired reads that aligned ≥1 times\n",
    "            result = subprocess.run(cmd, \n",
    "                                    check = True, \n",
    "                                    capture_output = True, \n",
    "                                    text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to align merged or unpaired fastq file {file.name}: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise\n",
    "        return result\n",
    "\n",
    "    def detect_reps(self, subfolder):\n",
    "        \"\"\"\n",
    "        defines variables for r1 and r2 filenames\n",
    "        \"\"\"\n",
    "        for r1_file in subfolder.glob(\"*unmerged_R1*\"):\n",
    "            self.r1_filename = r1_file\n",
    "            self.r2_filename = r1_file.with_name(r1_file.name.replace(\"_R1_\", \"_R2_\"))\n",
    "\n",
    "    def paired_reads(self, file):\n",
    "        \"\"\"\n",
    "        Align paired-end reads (unmerged)\n",
    "        \"\"\"\n",
    "        try: \n",
    "            cmd = [\"bowtie2\",\n",
    "                    \"-x\", str(self.bowtie2_index),\n",
    "                    \"-1\", str(self.r1_filename),\n",
    "                    \"-2\", str(self.r2_filename),\n",
    "                    \"-S\", str(self.sam_output), ## .sam file of contam reads\n",
    "                    \"--un-conc-gz\", str(self.rmcontam_output), ## unmerged reads that failed to align (mrna)\n",
    "                    \"--al-conc-gz\", str(self.contam_output)] ## unmerged reads that aligned ≥1 times\n",
    "            result = subprocess.run(cmd, \n",
    "                                    check = True, \n",
    "                                    capture_output = True, \n",
    "                                    text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to align unmerged fastq file {file.name}: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise\n",
    "        return result\n",
    "    \n",
    "    def convert_sam():\n",
    "        sam_filename = Path(self.sam_output).stem \n",
    "        bam_file = f\"{sam_filename}.bam\"\n",
    "\n",
    "        try:\n",
    "            subprocess.run([\"samtools\", \"sort\", \"-O\", \"BAM\", ## convert .sam to .bam and sort\n",
    "                            \"-o\", str(bam_file), str(self.sam_output)],\n",
    "                            check = True,\n",
    "                            capture_output = True,\n",
    "                            text = True)\n",
    "            subprocess.run([\"samtools\", \"index\", str(bam_file)], ## create .bai from .bam\n",
    "                            check = True,\n",
    "                            capture_output = True,\n",
    "                            text = True)\n",
    "            subprocess.run([\"rm\", str(self.sam_output)], ## remove original .sam file\n",
    "                            check = True,\n",
    "                            capture_output = True,\n",
    "                            text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to convert {sam_filename.name} to .bam file: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca57fc6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rmcontam_pipeline(folder_path, output_path):\n",
    "    if not os.path.exists(output_path):\n",
    "        os.makedirs(output_path) ## if output_path doesn't already exist, create new empty output folder\n",
    "\n",
    "    input_dir = Path(folder_path) ## imported Path to improve readability of code below\n",
    "    output_dir = Path(output_path)\n",
    "    aligner = Bowtie2Aligner(input_dir) ## initialize class\n",
    "    \n",
    "    aligner.build_bowtie2_index() ## build bowtie2 index\n",
    "\n",
    "    for subfolder in input_dir.iterdir(): ## amount of subfolders = number of replicates\n",
    "        if subfolder.is_dir():\n",
    "            processed_folder = output_dir/f\"{subfolder.name}_bowtie2\" ## processed folders for bowtie2 outputs\n",
    "            processed_folder.mkdir(exist_ok=True) ## if directory already exists, suppress OSError\n",
    "\n",
    "            for file in subfolder.glob(\"*.fastq.gz\"): ## iterate through indiv. files in subfolder\n",
    "                try:\n",
    "                    ## run bowtie2 alignment functions\n",
    "                    if \"_merged\" in file.name or \"_unpaired\" in file.name:\n",
    "                        aligner.single_reads(file, processed_folder)\n",
    "                    elif \"_unmerged\" in file.name:\n",
    "                        aligner.paired_reads(file, processed_folder)\n",
    "                    \n",
    "                    ## run samtools function\n",
    "                    aligner.convert_sam()\n",
    "                    \n",
    "                except Exception as e:\n",
    "                    print(f\"Failed to align {file.name} with bowtie2 and produce .bam files: {e}\")\n",
    "                    traceback.print_exc()\n",
    "                    continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23b1c93b",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser(description = \"Run contaminant removal pipeline.\")\n",
    "    parser.add_argument(\"--input\", required = True, help = \"Input processed_fastqs folder name\")\n",
    "    parser.add_argument(\"--output\", required = True, help = \"Desired output folder name for processed files\")\n",
    "    parser.add_argument(\"-u\", \"--unzip\", action = \"store_true\", help = \"Unzips fastq files\")\n",
    "\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    print(\"Starting contaminant removal pipeline...\")\n",
    "    rmcontam_pipeline(args.input, args.output)\n",
    "    print(\"Pipeline finished.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
